File: SDKs/ios/Sources/KISR/KISRRemote.swift
// ============================================================================
// EXAMPLE FILE
// ----------------------------------------------------------------------------
// Used as an example of how to bridge to your SDK/WASM
// ============================================================================
//
//
import Foundation

public final class KISRRemote {
	public struct Config {
		public let baseURL: URL
		public init(baseURL: URL) { self.baseURL = baseURL }
	}

	private let config: Config
	private let urlSession: URLSession

	public init(config: Config, session: URLSession = .shared) {
		self.config = config
		self.urlSession = session
	}

	// MARK: - Native Kaspa API fetch (payload only)
	private func normalizeNetwork(_ network: String) -> String {
		let n = network.lowercased()
		if n == "testnet" || n == "testnet-10" { return "testnet-10" }
		return "mainnet"
	}
	private func inferNetwork(from address: String?) -> String {
		guard let address else { return "mainnet" }
		if address.lowercased().hasPrefix("kaspatest:") { return "testnet-10" }
		return "mainnet"
	}
	private struct KaspaTxResponse: Decodable {
		let payload: String?
	}

	public func fetchPayload(txid: String, network: String = "mainnet", address: String? = nil) async throws -> String {
		let netInput = address != nil ? inferNetwork(from: address) : network
		let net = normalizeNetwork(netInput)
		let base = (net == "testnet-10") ? "https://api-tn10.kaspa.org" : "https://api.kaspa.org"
		var comps = URLComponents(string: base + "/transactions/" + txid)!
		comps.queryItems = [
			URLQueryItem(name: "inputs", value: "false"),
			URLQueryItem(name: "outputs", value: "false"),
			URLQueryItem(name: "resolve_previous_outpoints", value: "no")
		]
		guard let url = comps.url else { throw KISRError.payloadInvalid }
		var request = URLRequest(url: url)
		request.httpMethod = "GET"
		request.setValue("application/json", forHTTPHeaderField: "Accept")
		request.timeoutInterval = 8.0
		let (data, response) = try await urlSession.data(for: request)
		guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else { throw KISRError.payloadInvalid }
		let decoded = try JSONDecoder().decode(KaspaTxResponse.self, from: data)
		guard let payloadHex = decoded.payload, !payloadHex.isEmpty else { throw KISRError.payloadInvalid }
		return payloadHex
	}
}

File: SDKs/android/kisr/src/main/java/org/kisr/sdk/KISRRemote.kt
// ============================================================================
// KISR Modules (Multi-wallet support)
// ----------------------------------------------------------------------------
// This file is part of the KISR standard. Make sure to implement the exact encryption and code generations.
// ============================================================================
//
//
package org.kisr.sdk

import org.kisr.sdk.KISRError
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

class KISRRemote(private val config: Config = Config()) {
	data class Config(
		val mainnetBase: String = "https://api.kaspa.org",
		val testnetBase: String = "https://api-tn10.kaspa.org"
	)

	private fun normalizeNetwork(network: String): String {
		val n = network.lowercase()
		return if (n == "testnet" || n == "testnet-10") "testnet-10" else "mainnet"
	}

	fun fetchPayload(txid: String, network: String = "mainnet", address: String? = null): String {
		val net = normalizeNetwork(address?.let { if (it.lowercase().startsWith("kaspatest:")) "testnet-10" else network } ?: network)
		val base = if (net == "testnet-10") config.testnetBase else config.mainnetBase
		val url = URL("$base/transactions/$txid?inputs=false&outputs=false&resolve_previous_outpoints=no")
		val conn = (url.openConnection() as HttpURLConnection).apply {
			requestMethod = "GET"
			setRequestProperty("Accept", "application/json")
			connectTimeout = 8000
			readTimeout = 8000
		}
		conn.inputStream.use { stream ->
			val text = BufferedReader(InputStreamReader(stream)).readText()
			val payloadHex = parsePayloadField(text) ?: throw KISRError.PayloadInvalid()
			return payloadHex
		}
	}

	private fun parsePayloadField(json: String): String? {
		// Minimal parser to avoid bringing gson; expects { "payload": "hex" }
		val key = "\"payload\""
		val idx = json.indexOf(key)
		if (idx < 0) return null
		val colon = json.indexOf(':', idx + key.length)
		if (colon < 0) return null
		var i = colon + 1
		while (i < json.length && json[i].isWhitespace()) i++
		if (i >= json.length || json[i] != '"') return null
		i++
		val start = i
		while (i < json.length && json[i] != '"') i++
		if (i >= json.length) return null
		return json.substring(start, i)
	}
}
